### PR TITLE
Highlighting

### PR DESCRIPTION
**Pull request by vramesh1**
List of changes :

1) Result Factory separation for Knowledge Manager
- previously, Knowledge Manager and Custom searchers fell into a Generic Result Factory function
- creating a separate Result Factory function to create Results for Knowledge Manager backends
- refactor to move KM Result creation and Field Formatting logic into this function
2) Highlighting
- KM backends return highlighting instructions when available to highlight entity field values
- apply this highlighting to substrings within entity field values
- adding a property to the Result class to contain highlighted fields
- in addition to the property that contains the raw (without highlighting) entity profile data 
3) Highlighting available to Field Formatters
- pass in the highlighted field value to field formatters functions
- developers can choose to use either the raw entity field value or the highlighted version
4) Field Formatter interface
- modified the interface for Field Formatters to take in an object containing named formatter params
- instead of a growingly long list of parameters
5) Scope of Highlighter member function
- to consolidate logic to manipulate strings for highlighting
- re-using the HighlightedValue class' buildHighlightedValue function
- this was previously a private member function
- making it public for use by ResultFactory as well
6) Automated Tests
- added test suite for highlighting entity profile data
- changed field formatting test suite to use the modified field formatter interface

### CODE REVIEW

File: src/core/models/highlightedvalue.js
Comment (b1lly) [inline src/core/models/highlightedvalue.js:77]: Can you add a comment on top of the function to describe what the structure of these values look like?

You can use JSDoc to help communicate that, and/or also give a example object.

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js]: Style nit: can we name `data` = `entityProfile` or `entityProfileData`?

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js]: style nit: to avoid all the indentations can we just short circuit?

```
if (Object.keys(formatters.length) === 0) {
  return formattedData;
}

Object.entries(data).forEach([....]) => {
  // This can if (formatters[fieldName] === undefined) { ... }
  // Or, you can explicitly ensure that it's a function (e.g. below)...
  // It just depends on how safe you want to be and how you want to handle exceptions
  if (typeof formatters[fieldName] !== 'function') {
    // This skips the current iteration
    return;
  }

  let highlightedFieldObject = ResultFactory.getHighlightedData(profileData, highlightedField);
  ...
}

return formattedData;
```

**Review by b1lly — COMMENTED** [review]
Nice dude! Minor style nits.

**Comment by Delerme** [comment]
this'll need a rebase on a current version of v0.9.0 before review

**Comment by vramesh1** [comment]
@b1lly @Delerme , changes made per comments and ready for review when you get a chance. thanks.

File: src/core/models/highlightedvalue.js
Comment (b1lly) [inline src/core/models/highlightedvalue.js]: JSDoc is like this:

```
@param name {type} description
```

e.g.

```
@param highlightedSubstrings {object} highlighting specifies to apply to input object
@returns {string} The highlighted value
```

At some point, we use a type checking compiler, but in the short term, at least it gives the developers a little more context into the structure of the object.

Your example is amazing btw.

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js]: This is almost like swallowing an exception (the typechecking case). Just wanted to make sure you were explicit about what behavior you wanted here.

For instance, if someone did:
```
fieldFormatters: {
  'name': "NERD"
};
```
In your case, it would just ignore it (which is fine). The alternative is to explicitly throw an AnswersError that tells them they did something bad.

I don't feel strongly either way, was just making sure the behavior was intended.

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js]: Style nit: If it's an annoying refactor, ignore this comment, but typically I tend to avoid `getName` methods unless it's an explicit read only operation. For instance, in this case, i'd probably name it "computeHighlightData" since the operation is actually doing work.

Reasoning: A lot of people fall into the trap of overloading the `get[Name]` practice, which can cause others to use the methods improperly.

For instance, i've seen a lot of cases where people implement a `getMethod` which turns out to be an `expensive` operation, and then other developers end up using it in a tight loop without realizing the operation was actually doing something heavy.

Because of that, I try to be as explicit as I can with my naming... to help others avoid making mistakes.

I know you didn't name the original, but wanted to give some context.

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js]: Nit: You could probably just inline this into the consumer, instead of storing a reference.

```
formattedData[fieldName] = formatters[fieldName]({
  ...
});
```

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js]: computeHighlightedData ?

**Review by b1lly — CHANGES_REQUESTED** [review]
You're a savage! Mostly style nits.

Sorry I didn't catch these in the first iteration, but this looks pretty solid.

File: src/core/models/highlightedvalue.js
Comment (vramesh1) [reply src/core/models/highlightedvalue.js]: Sure. Adding Type.
Looks like it's actually 

> @param type name

instead of the other way around.
Not sure if that matters anyway for now.

File: src/core/models/resultfactory.js
Comment (vramesh1) [reply src/core/models/resultfactory.js]: sure, adding exception handling for a couple of cases across formatting and highlighting functions

File: src/core/models/highlightedvalue.js
Comment (b1lly) [reply src/core/models/highlightedvalue.js]: yeah, woops!

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js]: Woops, don't we want to use the highlighted data here by default, if it exists?

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js]: what made you decide to process the formatting here, instead of within the factory? but you process the highlighted data within the factory?

Also, I think you're computing the highlighted data twice.

Perhaps you should move both computes into the factory, but do the highlighting first, and then pipe the results into the formatter.

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js]: Maybe better: "Field formatter is not of expected type `function`"

Just a little more descriptive.

File: src/core/models/resultfactory.js
Comment (b1lly) [inline src/core/models/resultfactory.js:128]: See comment above.

**Review by b1lly — CHANGES_REQUESTED** [review]

**Review by Delerme — COMMENTED** [review]
Looks good, will let billy approve after the minor changes

**Comment by b1lly** [comment]
I think we need to rebase this before we can merge it back in.

Should we rebase this onto the v0.9.2 branch or should this be part of 1.0.0.
@Delerme thoughts?

**Comment by Delerme** [comment]
Up to you all, I haven't heard anything about 1.0.0 so am not sure what the plan is there

**Comment by vramesh1** [comment]
Created a branch for v1.0.0 and merged v0.9.2 changes into it.
Updated the base branch of this PR to v1.0.0.

**Comment by vramesh1** [comment]
created a separate PR to merge these changes into new base branch v1.0.0.
https://github.com/yext/answers/pull/325

**Comment by vramesh1** [comment]
closing in favour of 
https://github.com/yext/answers/pull/325

**Pull request closed without merging**
